src/test/java/edu/hm/hafner/util/SetTest.java

```diff
@@ -0,0 +1,179 @@
+    }
+
+    @Test
+    void isEmpty() {
+        testInitializing();
+
+        //Test: Creating a new variable
+        assertThat(testHash.isEmpty()).isTrue();
+
+        //Test: Adding values, so the hash is not empty any more.
+        testHash.add(2503);
+        assertThat(testHash.isEmpty()).isFalse();
+        testHash.add(1106);
+        assertThat(testHash.isEmpty()).isFalse();
+        testHash.add(2212);
+        assertThat(testHash.isEmpty()).isFalse();
+        testHash.add(709);
+        assertThat(testHash.isEmpty()).isFalse();
+
+        //Test: Removing all values in the hash step by step.
+        testHash.remove(2503);
+        assertThat(testHash.isEmpty()).isFalse();
+        testHash.remove(1106);
+        assertThat(testHash.isEmpty()).isFalse();
+        testHash.remove(2212);
+        assertThat(testHash.isEmpty()).isFalse();
+        testHash.remove(709);
+        assertThat(testHash.isEmpty()).isTrue();
+
+        // Test: Adding and removing more than one Object.
+        testHash.addAll(extendHash);
+        assertThat(testHash.isEmpty()).isFalse();
+        testHash.removeAll(extendHash);
+        assertThat(testHash.isEmpty()).isTrue();
+    }
+
+    @Test
+    void contains() {
+        testInitializing();
+        assertThat(extendHash.contains(1106)).isTrue();
+        assertThat(extendHash.contains(2503)).isTrue();
+        assertThat(extendHash.contains(2212)).isTrue();
+        assertThat(extendHash.contains(709)).isTrue();
+        assertThat(extendHash.contains(4)).isFalse();
+        assertThat(extendHash.contains(0)).isFalse();
+        assertThat(extendHash.contains(1999)).isFalse();
+    }
+
+    @Test
+    void add() {
+        testHash = new HashSet<>();
+        assertThat(testHash).isEmpty();
+        testHash.add(50);
+        assertThat(testHash.hashCode()).isEqualTo(50);
+        testHash.add(443);
+        assertThat(testHash.hashCode()).isEqualTo(493); //50+443
+    }
+
+    @Test
+    void remove() {
+        testInitializing();
+        assertThat(extendHash).isNotNull();
+        assertThat(extendHash.hashCode()).isEqualTo(6530); //All extendedHash integers summed
+        extendHash.remove(709);
+        assertThat(extendHash.hashCode()).isEqualTo(5821); //All extendedHash integers summed without 709
+        extendHash.remove(1106);
+        extendHash.remove(2503);
+        extendHash.remove(2212);
+        assertThat(extendHash.hashCode()).isEqualTo(0);
+        assertThat(extendHash).isEmpty();
+    }
+
+    @Test
+    void clear() {
+        testInitializing();
+        extendHash.clear();
+        assertThat(extendHash).isEmpty();
+    }
+
+    /**
+     * Test variables initializing.
+     */
+    void testInitializing() {
+        testHash = new HashSet<>(); //empty
+        extendHash = new HashSet<>();
+        extendHash.add(1106);
+        extendHash.add(2503);
+        extendHash.add(2212);
+        extendHash.add(709);
+    }
+}
```

Comment:
Die clone-Methode wird nicht getestet.

Reply:
Bei mir kommt bei der clone()-Methode folgende Warnung:
'clone()' in 'edu.hm.hafner.util.SetTest' clashes with 'clone()' in 'java.lang.Object'; attempting to assign weaker access privileges ('package-private'); was 'protected'

Reply:
Du hast die Testmethode scheinbar 'clone' genannt. Wie du vielleicht weißt, erbt jede Klasse von der Klasse `Object`. Dort ist eine Methode `clone` definiert, die du dadurch überschreibst. Beim Überschreiben einer Methode darf der Zugriff durch den verwendeten access modifier nur gelockert aber nicht eingeschränkt werden. Deshalb kommt diese Fehlermeldung. Da du die Methode `clone` in der Klasse `Object` aber gar nicht überschreiben willst, verwende einfach einen anderen Methodennamen. Am besten bei Namen für Testmethoden immer einen Prefix wie 'should' oder 'test' verwenden. Also zum Beispiel 'shouldClone'.